### PR TITLE
Access Control: Add service accounts

### PR DESCRIPTION
**Pull request by donomii**
**What this PR does / why we need it**:

This PR integrates code that will be the basis for further work on service accounts.  It is an interim release to prevent merge conflicts and to show the direction we are taking to implement service accounts.

**Which issue(s) this PR fixes**:

Fixes https://github.com/grafana/grafana-enterprise/issues/1858

**Special notes for your reviewer**:

This is partial release to integrate the current work on service accounts with main.  The patch was growing too large to manage, so we are breaking the work into smaller releases.

This PR introduces the basic components needed to implement service accounts.  It changes the user and api_key tables, and the structures needed to deal with them.

It also modifies the "create apikey" api call, allowing for some basic testing of the entire patch.

The goal of this patch is not to add functionality, but to integrate our current work and prevent merge conflicts later, as well as showing the direction we are taking in implementing service accounts.

### CODE REVIEW

File: public/app/features/api-keys/ApiKeysForm.tsx [hunk 1]
```diff
@@ -99,6 +119,17 @@ export const ApiKeysForm: FC<Props> = ({ show, onClose, onKeyAdded }) => {
...
             </div>
+            <div className="gf-form">
+              <label>
+                New Service Account
+                <input
+                  name="createServiceAccount"
+                  type="checkbox"
+                  checked={createServiceAccount}
+                  onChange={onCreateServiceAccountChange}
+                />
```
Comment (hugohaggmark) [inline public/app/features/api-keys/ApiKeysForm.tsx:130]: nit: maybe change to more explicit boolean
```suggestion
                <input
                  name="createServiceAccount"
                  type="checkbox"
                  checked={Boolean(createServiceAccount)}
                  onChange={onCreateServiceAccountChange}
                />
```

File: public/app/features/api-keys/ApiKeysPage.tsx [hunk 2]
```diff
@@ -76,6 +76,7 @@ export class ApiKeysPageUnconnected extends PureComponent<Props, State> {
   };
 
   onAddApiKey = (newApiKey: NewApiKey) => {
+    console.log(newApiKey);
```
Comment (hugohaggmark) [inline public/app/features/api-keys/ApiKeysPage.tsx:79]: Please remove

**Review by hugohaggmark — DISMISSED** [review]
Left some comments about the frontend parts and also the UI needs more polish. I didn't see the `New Service Account` checkbox 
image

**Comment by kminehart** [comment]
@donomii I'd be happy to review this! I think a lot of what I'm looking at (in the backend) has a lot of "todo" markers. Since this is WIP, when you're ready for a review on the Go stuff, feel free to ping me!

**Comment by donomii** [comment]
Thanks!

I made a mistake with Github and didn’t set the PR to draft, so it added a lot of reviewers to the PR.  Sorry about that!

It will be a couple of weeks before it’s ready for review, I’ll let everyone know when it is finished.

**Review by eleijonmarck — COMMENTED** [review]
just one missing err I think

**Review by eleijonmarck — DISMISSED** [review]
accounts

**Review by sakjur — DISMISSED** [review]
I _think_ this breaks API keys from the perspective of the OSS codebase. Let's hold off on merging until that's cleared out.

I'm also a bit concerned about introducing panics, we have error handling in place in each of the places where panics are being introduced and I don't think any of them should be considered runtime errors, I think we should avoid panicking because we _can_ effectively recover from these errors. (see [effective go on panics](https://golang.org/doc/effective_go#panic))

**Review by IevaVasiljeva — COMMENTED** [review]
Some questions and suggestions in comments. Feel free to ignore the nits if they don't align with your personal preferences 🙂 

I'd also love to test the changes out - what's the best way to do it? Should I call the endpoints directly or is there a nice way to test them through the UI?

**Comment by donomii** [comment]
> Some questions and suggestions in comments. Feel free to ignore the nits if they don't align with your personal preferences 🙂
> 
> I'd also love to test the changes out - what's the best way to do it? Should I call the endpoints directly or is there a nice way to test them through the UI?

I removed the UI for now, but I can add that back into another branch if you are interested.  For now you can call the API and set the CreateNewServiceAccount flag to true.  You can then look at the database table to see the new service account in the apikey table

**Comment by donomii** [comment]
pkg/models/user.go
	Result           User
	IsServiceAccount bool
 
Member
@IevaVasiljeva IevaVasiljeva 23 hours ago
nit: is Result User separated from the rest of the fields because it is the result of the command? If so, should IsServiceAccount bool be defined together with the rest of the attributes higher up?

This is a left over habit I have from another programming language, where it is very important to always add new fields at the end of the structure.  But because we have a monolithic program, we can add them anywhere.

**Review by marefr — COMMENTED** [review]
Added some comments/questions. In general looks good, but feels like you've might added to much code in certain places that doesn't do anything and could affect existing functionality - please make sure to guard everywhere with the feature toggle if you want to keep things as is. 

Have #16424 been considered in relation to this?

**Review by gamab — COMMENTED** [review]
I just started looking at your work, so this is unrelated with the core work but I found some mismatches in the mock :slightly_smiling_face:

**Review by eleijonmarck — APPROVED** [review]
LGTM! looks amazing

**Pull request merged**